std/algorithm/searching.d

```diff
@@ -119,11 +119,15 @@ template all(alias pred = "a")
     Performs (at most) $(BIGOH range.length) evaluations of `pred`.
      +/
     bool all(Range)(Range range)
-    if (isInputRange!Range &&
-        (__traits(isTemplate, pred) || is(typeof(unaryFun!pred(range.front)))))
     {
         import std.functional : not;
 
+        static assert(isInputRange!Range, Range.stringof, " must be an"
+                , " InputRange");
+        static assert(__traits(isTemplate, pred)
+                    || is(typeof(unaryFun!pred(range.front)))
+                , "The passed predicate (pred), must be usable by unaryFun");
```

Comment:
This could be clearer - it also tests that `range.front` can be passed to `unaryFun`. Though it might be better not to have a static assert for this at all (see above).

Reply:
Sorry, I removed my other comments which don't apply to this pull as the `isTemplate` check is kept.

However, when `pred` is not a template, using a static assert hides a more descriptive error when the parameter is the wrong type. When `pred` is `(string s) => s.length` and `range` is `int[]`, I get these error without constraints or static assert:
```
/usr/include/dmd/phobos/std/functional.d(642): Error: function literal `__lambda2(string s)` is not callable using argument types `(int)`
/usr/include/dmd/phobos/std/functional.d(642):        cannot pass argument `_param_0` of type `int` to parameter `string s`
/usr/include/dmd/phobos/std/algorithm/searching.d(1869): Error: template instance `std.functional.not!(function (string s) pure nothrow @nogc @safe => s.length).not!int` error instantiating
find.d(11):        instantiated from here: `find!(not, int[])`
```